<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>CopyKit Workflow â€¢ CopyKit</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script>




<meta property="og:title" content="CopyKit Workflow" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-title-body">
      <header>
      <div class="navbar navbar-invert navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">CopyKit</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9005</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>CopyKit Workflow</h1>
    </div>


<div id="copykit-workflow" class="section level1">

<p><strong>CopyKit</strong> provides a toolkit for the analysis of single-cell copy number datasets. It includes functions to read data from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4417119/">Varbin</a> or <a href="https://www.10xgenomics.com/solutions/single-cell-cnv/">10X CNA datasets</a>.</p>
<p>A common workflow with <strong>CopyKit</strong> consists in reading the dataset and using to filter the noisy cells out, clustering and plotting heatmaps for further analysis.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install the development version of CopyKit from github with:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">devtools<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/devtools/man/reexports.html">install_github</a></span>(<span class="st">"navinlabcode/copykit"</span>)</a></code></pre></div>
</div>
</div>
<div id="getting-started" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting Started</h1>
<p><strong>CopyKit</strong> provides a toolkit for the analysis of single-cell copy number datasets. It includes functions to read data from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4417119/">Varbin</a> or <a href="https://www.10xgenomics.com/solutions/single-cell-cnv/">10X CNA datasets</a>.</p>
<p>A common workflow with <strong>CopyKit</strong> consists in reading the dataset and using to filter the noisy cells out, clustering and plotting heatmaps for further analysis.</p>
<div id="reading-varbin-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-varbin-dataset" class="anchor"></a>Reading Varbin dataset</h2>
<p>Datasets generated with the varbin pipeline can be read with the function <code><a href="reference/readVarbinCNA.html">readVarbinCNA()</a></code>. It creates an S4 class <code>scCNA</code> from the output directory of the copy number pipeline. <code><a href="reference/readVarbinCNA.html">readVarbinCNA()</a></code> searches for the uber*.seg uber.bin and uber.ratio files in the provided directory.</p>
<p>The scCNA object will contain 3 different assays: <strong>segment ratios</strong>, <strong>ratios</strong> and <strong>bincounts</strong> where each bin is row and each sample (cell) is a column.</p>
<p>CopyKit stores the genomic ranges as a GRanges object containing chromosome number, start coordinate, end coordinate and absolute genomic position. Each row represents the coordinates for one bin.</p>
<p>It is possible to exclude the information from ChrY (important when working with XX samples) by setting the argument <code>remove_Y</code> to <code>TRUE</code>. Ex: <code><a href="reference/readVarbinCNA.html">readVarbinCNA("~/path_to_varbin_output/", remove_Y = TRUE)</a></code>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">breast_tumor &lt;-<span class="st"> </span>copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/readVarbinCNA.html">readVarbinCNA</a></span>(<span class="st">"~/path_to_varbin_output/"</span>)</a></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co">## Importing segment ratios.</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="co">## Importing ratios.</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="co">## Importing bin counts.</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="co">## Removed ChrY information.</span></a></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">breast_tumor</a></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co">## class: scCNA </span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="co">## dim: 12167 1442 </span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="co">## metadata(0):</span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="co">## assays(3): segment_ratios ratios bin_counts</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="co">## rownames(12167): 1 2 ... 12166 12167</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="co">## rowData names(6): bin_length gene_count ... gc_content abspos</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="co">## colnames(1442): tn20_2_s2_c10_s394_r1_001 tn20_2_s2_c11_s395_r1_001 ...</span></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="co">##   tn20_s6_c382_s2302_r1_001 tn20_s6_c383_s2303_r1_001</span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="co">## colData names(1): sample</span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10"><span class="co">## reducedDimNames(0):</span></a>
<a class="sourceLine" id="cb5-11" data-line-number="11"><span class="co">## spikeNames(0):</span></a>
<a class="sourceLine" id="cb5-12" data-line-number="12"><span class="co">## rowRanges has: 12167 ranges</span></a>
<a class="sourceLine" id="cb5-13" data-line-number="13"><span class="co">## Phylo:</span></a></code></pre></div>
<p><strong>CopyKit</strong> stores the bin information within a rowRanges object, which can be accessed with <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">SummarizedExperiment::rowRanges()</a></code>:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">SummarizedExperiment<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">rowRanges</a></span>(breast_tumor)</a></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="co">## GRanges object with 12167 ranges and 6 metadata columns:</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="co">##         seqnames              ranges strand | bin_length gene_count cgi_count</span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="co">##            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;integer&gt;  &lt;integer&gt; &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co">##       1     chr1      977837-1200863      * |     223027         11        41</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="co">##       2     chr1     1200864-1455238      * |     254375         18        33</span></a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="co">##       3     chr1     1455239-1758057      * |     302819         11        35</span></a>
<a class="sourceLine" id="cb7-7" data-line-number="7"><span class="co">##       4     chr1     1758058-1989903      * |     231846          5        19</span></a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="co">##       5     chr1     1989904-2211810      * |     221907          3        16</span></a>
<a class="sourceLine" id="cb7-9" data-line-number="9"><span class="co">##     ...      ...                 ...    ... .        ...        ...       ...</span></a>
<a class="sourceLine" id="cb7-10" data-line-number="10"><span class="co">##   12163     chrX 153835263-154107535      * |     272273          8         4</span></a>
<a class="sourceLine" id="cb7-11" data-line-number="11"><span class="co">##   12164     chrX 154107536-154339310      * |     231775         12         4</span></a>
<a class="sourceLine" id="cb7-12" data-line-number="12"><span class="co">##   12165     chrX 154339311-154639929      * |     300619          4         3</span></a>
<a class="sourceLine" id="cb7-13" data-line-number="13"><span class="co">##   12166     chrX 154639930-154938160      * |     298231          1         3</span></a>
<a class="sourceLine" id="cb7-14" data-line-number="14"><span class="co">##   12167     chrX 154938161-155180157      * |     241997          2         1</span></a>
<a class="sourceLine" id="cb7-15" data-line-number="15"><span class="co">##         dist_telomere        gc_content      abspos</span></a>
<a class="sourceLine" id="cb7-16" data-line-number="16"><span class="co">##             &lt;integer&gt;         &lt;numeric&gt; &lt;integer64&gt;</span></a>
<a class="sourceLine" id="cb7-17" data-line-number="17"><span class="co">##       1       1146783 0.611899007743492      977837</span></a>
<a class="sourceLine" id="cb7-18" data-line-number="18"><span class="co">##       2       1371964 0.592286977886868     1200864</span></a>
<a class="sourceLine" id="cb7-19" data-line-number="19"><span class="co">##       3       1699329 0.545698915854107     1455239</span></a>
<a class="sourceLine" id="cb7-20" data-line-number="20"><span class="co">##       4       1934728 0.541937320462544     1758058</span></a>
<a class="sourceLine" id="cb7-21" data-line-number="21"><span class="co">##       5       2156893 0.570018070633446     1989904</span></a>
<a class="sourceLine" id="cb7-22" data-line-number="22"><span class="co">##     ...           ...               ...         ...</span></a>
<a class="sourceLine" id="cb7-23" data-line-number="23"><span class="co">##   12163       1227655 0.433821936071646  3034868549</span></a>
<a class="sourceLine" id="cb7-24" data-line-number="24"><span class="co">##   12164        983007 0.395383453780725  3035140822</span></a>
<a class="sourceLine" id="cb7-25" data-line-number="25"><span class="co">##   12165        749886 0.390856865334616  3035372597</span></a>
<a class="sourceLine" id="cb7-26" data-line-number="26"><span class="co">##   12166        391320 0.380503703505117  3035673216</span></a>
<a class="sourceLine" id="cb7-27" data-line-number="27"><span class="co">##   12167        155540 0.383103922776009  3035971447</span></a>
<a class="sourceLine" id="cb7-28" data-line-number="28"><span class="co">##   -------</span></a>
<a class="sourceLine" id="cb7-29" data-line-number="29"><span class="co">##   seqinfo: 24 sequences from an unspecified genome; no seqlengths</span></a></code></pre></div>
</div>
<div id="filtering-cells-" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-cells-" class="anchor"></a>Filtering cells.</h2>
<p>Cells with low quality can be easily filtered with <code><a href="reference/filterCells.html">filterCells()</a></code>.</p>
<p><code><a href="reference/filterCells.html">filterCells()</a></code> uses a k-nearest-neighbor approach to remove cells with random CNA profiles, largely due to noise.</p>
<p>By default, <strong>CopyKit</strong> automatically outputs a heatmap so you can visually assess if the filtering performed as expected, this however can be turned off by adding the argument <code><a href="reference/filterCells.html">filterCells(scCNA, plot_heatmap = FALSE)</a></code>. Plotting the heatmap is a time consuming step due to distance calculations. CopyKit also allows for parallel processing of the distance calculations <code><a href="reference/filterCells.html">filterCells()</a></code> with the argument <code>n_threads</code>.</p>
<p>To control how strict the filtering, change the argument <code>resolution</code>. Ex: <code><a href="reference/filterCells.html">filterCells(resolution = 0.9)</a></code></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">breast_tumor &lt;-<span class="st"> </span>copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/filterCells.html">filterCells</a></span>(breast_tumor,</a>
<a class="sourceLine" id="cb8-2" data-line-number="2">                                     <span class="dt">resolution =</span> <span class="fl">0.8</span>)</a></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="co">## Calculating correlation matrix.</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2"></a>
<a class="sourceLine" id="cb9-3" data-line-number="3"><span class="co">## Adding information to metadata. Access with SummarizedExperiment::colData(scCNA).</span></a>
<a class="sourceLine" id="cb9-4" data-line-number="4"></a>
<a class="sourceLine" id="cb9-5" data-line-number="5"><span class="co">## Plotting heatmap.</span></a>
<a class="sourceLine" id="cb9-6" data-line-number="6"></a>
<a class="sourceLine" id="cb9-7" data-line-number="7"><span class="co">## Your dataset has: 1442 Cells. Plotting heatmap may take a long time with large number of cells. Set number of threads with n_threads for parallel processing if possible to speed up.</span></a>
<a class="sourceLine" id="cb9-8" data-line-number="8"></a>
<a class="sourceLine" id="cb9-9" data-line-number="9"><span class="co">## Done.</span></a></code></pre></div>
<p><img src="copykit_workflow_files/figure-gfm/filtering_cells_parallel-1.png"><!-- --></p>
<p>The information is store as metadata and can be accessed with <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">SummarizedExperiment::colData()</a></code>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(SummarizedExperiment<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span>(breast_tumor))</a></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co">## DataFrame with 6 rows and 2 columns</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="co">##                                              sample    filtered</span></a>
<a class="sourceLine" id="cb11-3" data-line-number="3"><span class="co">##                                         &lt;character&gt; &lt;character&gt;</span></a>
<a class="sourceLine" id="cb11-4" data-line-number="4"><span class="co">## tn20_2_s2_c10_s394_r1_001 tn20_2_s2_c10_s394_r1_001        kept</span></a>
<a class="sourceLine" id="cb11-5" data-line-number="5"><span class="co">## tn20_2_s2_c11_s395_r1_001 tn20_2_s2_c11_s395_r1_001        kept</span></a>
<a class="sourceLine" id="cb11-6" data-line-number="6"><span class="co">## tn20_2_s2_c12_s396_r1_001 tn20_2_s2_c12_s396_r1_001     removed</span></a>
<a class="sourceLine" id="cb11-7" data-line-number="7"><span class="co">## tn20_2_s2_c13_s397_r1_001 tn20_2_s2_c13_s397_r1_001     removed</span></a>
<a class="sourceLine" id="cb11-8" data-line-number="8"><span class="co">## tn20_2_s2_c14_s398_r1_001 tn20_2_s2_c14_s398_r1_001     removed</span></a>
<a class="sourceLine" id="cb11-9" data-line-number="9"><span class="co">## tn20_2_s2_c19_s403_r1_001 tn20_2_s2_c19_s403_r1_001        kept</span></a></code></pre></div>
<p>If you are satisfied with the filtering you can simply subset the object based on the <code>SummarizedExperiment::colData(bt)$filtered</code> column</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">breast_tumor &lt;-<span class="st"> </span>breast_tumor[,SummarizedExperiment<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span>(breast_tumor)<span class="op">$</span>filtered <span class="op">==</span><span class="st"> "kept"</span>]</a></code></pre></div>
<p>After subsetting it is recommended to calculate a distance matrix that will be later passaged to functions such as <code><a href="reference/runPhylo.html">runPhylo()</a></code> or <code><a href="reference/plotHeatmap.html">plotHeatmap()</a></code>. This distance matrix is stored in the scCNA object and can be accessed with <code>distMat()</code>.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">breast_tumor &lt;-<span class="st"> </span>copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/runDistMat.html">runDistMat</a></span>(breast_tumor)</a></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="co">## Calculating distance matrix with metric: euclidean</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2"></a>
<a class="sourceLine" id="cb14-3" data-line-number="3"><span class="co">## Using 48 cores.</span></a>
<a class="sourceLine" id="cb14-4" data-line-number="4"></a>
<a class="sourceLine" id="cb14-5" data-line-number="5"><span class="co">## Access distance matrix with copykit::distMat()</span></a>
<a class="sourceLine" id="cb14-6" data-line-number="6"></a>
<a class="sourceLine" id="cb14-7" data-line-number="7"><span class="co">## Done.</span></a></code></pre></div>
</div>
<div id="non-linear-dimension-reduction" class="section level2">
<h2 class="hasAnchor">
<a href="#non-linear-dimension-reduction" class="anchor"></a>Non-linear dimension reduction</h2>
<p>We can explore the sample organization at reduced dimensions with <code><a href="reference/runUmap.html">runUmap()</a></code>. The goal is to explore the dataset for the presence of different subpopulations. <code><a href="reference/runUmap.html">runUmap()</a></code> stores the generated embedding into the <code>scCNA</code> object and can be accessed with <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SCE-reducedDims.html">SingleCellExperiment::reducedDim(scCNA, 'umap', withDimnames = FALSE)</a></code>. <strong>CopyKit</strong> uses the package <code>uwot</code> to generate the umap embedding, this means that additional parameters that control the global or local structure to <code><a href="reference/runUmap.html">runUmap()</a></code> can be passed on with the <code>...</code> argument. Ex: <code><a href="reference/runUmap.html">runUmap(scCNA, min_dist = 0)</a></code>. To learn more on how those parameters can control global and local structure embedding visit the <a href="https://umap-learn.readthedocs.io/en/latest/clustering.html">Umap website</a>. For the full list of possible parameters check <code><a href="https://rdrr.io/pkg/uwot/man/umap.html">?uwot::umap</a></code>.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">breast_tumor &lt;-<span class="st"> </span>copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/runUmap.html">runUmap</a></span>(breast_tumor,</a>
<a class="sourceLine" id="cb15-2" data-line-number="2">                                 <span class="dt">min_dist =</span> <span class="dv">0</span>,</a>
<a class="sourceLine" id="cb15-3" data-line-number="3">                                 <span class="dt">n_neighbors =</span> <span class="dv">30</span>)</a></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="co">## Embedding data with UMAP. Using seed 17</span></a>
<a class="sourceLine" id="cb16-2" data-line-number="2"></a>
<a class="sourceLine" id="cb16-3" data-line-number="3"><span class="co">## Access reduced dimensions slot with: SingleCellExperiment::reducedDim(scCNA, 'umap').</span></a>
<a class="sourceLine" id="cb16-4" data-line-number="4"></a>
<a class="sourceLine" id="cb16-5" data-line-number="5"><span class="co">## Done.</span></a></code></pre></div>
<p>It is possivel to visualize the results with the function <code><a href="reference/plotUmap.html">plotUmap()</a></code>.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw"><a href="reference/plotUmap.html">plotUmap</a></span>(breast_tumor)</a></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="co">## No cluster information detected, use findClusters() to create it.</span></a>
<a class="sourceLine" id="cb18-2" data-line-number="2"></a>
<a class="sourceLine" id="cb18-3" data-line-number="3"><span class="co">## Plotting Umap.</span></a></code></pre></div>
<p><img src="copykit_workflow_files/figure-gfm/unnamed-chunk-6-1.png"><!-- --></p>
</div>
</div>
<div id="finding-clusters" class="section level1">
<h1 class="hasAnchor">
<a href="#finding-clusters" class="anchor"></a>Finding clusters</h1>
<p><strong>CopyKit</strong> uses a graph based approacht to search for clusters in the scCNA dataset. <code><a href="reference/findClusters.html">findClusters()</a></code> builds an SNN graph of the k-nearest neighbors and attempts to find two different level of clustering: Major and minor subpopulations. Major clusters are found by looking at the graph connected components, whereas minor clusters uses the Leiden algorithm to detect connected communities within the major clusters. To run <code><a href="reference/findClusters.html">findClusters()</a></code> you first need <code><a href="reference/runUmap.html">runUmap()</a></code>. <strong>CopyKit</strong> applies the clustering on top of the UMAP embedding. Check <code>?copykit::findClusters()</code> for help on how to change the clustering parameters arguments.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">breast_tumor &lt;-<span class="st"> </span>copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/findClusters.html">findClusters</a></span>(breast_tumor)</a></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="co">## Building SNN graph.</span></a>
<a class="sourceLine" id="cb20-2" data-line-number="2"></a>
<a class="sourceLine" id="cb20-3" data-line-number="3"><span class="co">## Finding clusters.</span></a>
<a class="sourceLine" id="cb20-4" data-line-number="4"></a>
<a class="sourceLine" id="cb20-5" data-line-number="5"><span class="co">## Done.</span></a></code></pre></div>
<p>Clustering information is stored as metadata and can be accessed with `SummarizedExperiment::colData(scCNA).</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(SummarizedExperiment<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span>(breast_tumor))</a></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="co">## DataFrame with 6 rows and 4 columns</span></a>
<a class="sourceLine" id="cb22-2" data-line-number="2"><span class="co">##                                              sample    filtered major_clusters</span></a>
<a class="sourceLine" id="cb22-3" data-line-number="3"><span class="co">##                                         &lt;character&gt; &lt;character&gt;    &lt;character&gt;</span></a>
<a class="sourceLine" id="cb22-4" data-line-number="4"><span class="co">## tn20_2_s2_c10_s394_r1_001 tn20_2_s2_c10_s394_r1_001        kept              A</span></a>
<a class="sourceLine" id="cb22-5" data-line-number="5"><span class="co">## tn20_2_s2_c11_s395_r1_001 tn20_2_s2_c11_s395_r1_001        kept              A</span></a>
<a class="sourceLine" id="cb22-6" data-line-number="6"><span class="co">## tn20_2_s2_c19_s403_r1_001 tn20_2_s2_c19_s403_r1_001        kept              A</span></a>
<a class="sourceLine" id="cb22-7" data-line-number="7"><span class="co">## tn20_2_s2_c2_s386_r1_001   tn20_2_s2_c2_s386_r1_001        kept              B</span></a>
<a class="sourceLine" id="cb22-8" data-line-number="8"><span class="co">## tn20_2_s2_c20_s404_r1_001 tn20_2_s2_c20_s404_r1_001        kept              B</span></a>
<a class="sourceLine" id="cb22-9" data-line-number="9"><span class="co">## tn20_2_s2_c21_s405_r1_001 tn20_2_s2_c21_s405_r1_001        kept              A</span></a>
<a class="sourceLine" id="cb22-10" data-line-number="10"><span class="co">##                           minor_clusters</span></a>
<a class="sourceLine" id="cb22-11" data-line-number="11"><span class="co">##                                &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb22-12" data-line-number="12"><span class="co">## tn20_2_s2_c10_s394_r1_001              3</span></a>
<a class="sourceLine" id="cb22-13" data-line-number="13"><span class="co">## tn20_2_s2_c11_s395_r1_001              3</span></a>
<a class="sourceLine" id="cb22-14" data-line-number="14"><span class="co">## tn20_2_s2_c19_s403_r1_001             12</span></a>
<a class="sourceLine" id="cb22-15" data-line-number="15"><span class="co">## tn20_2_s2_c2_s386_r1_001               8</span></a>
<a class="sourceLine" id="cb22-16" data-line-number="16"><span class="co">## tn20_2_s2_c20_s404_r1_001              8</span></a>
<a class="sourceLine" id="cb22-17" data-line-number="17"><span class="co">## tn20_2_s2_c21_s405_r1_001              3</span></a></code></pre></div>
<p><code><a href="reference/plotUmap.html">plotUmap()</a></code> will display cluster information if it is available.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="kw"><a href="reference/plotUmap.html">plotUmap</a></span>(breast_tumor)</a></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="co">## Plotting Umap.</span></a>
<a class="sourceLine" id="cb24-2" data-line-number="2"></a>
<a class="sourceLine" id="cb24-3" data-line-number="3"><span class="co">## Using colData(scCNA) cluster information.</span></a></code></pre></div>
<p><img src="copykit_workflow_files/figure-gfm/unnamed-chunk-9-1.png"><!-- --></p>
</div>
<div id="running-phylogenetic-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#running-phylogenetic-analysis" class="anchor"></a>Running phylogenetic analysis</h1>
<p>It is possible to run phylogenetic analysis with copykit by using the function <code><a href="reference/runPhylo.html">runPhylo()</a></code>. This function creates a distance based neighbor-joining tree that is stored within the scCNA object. You can access the resulting tree with <code>phylo()</code></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1">breast_tumor &lt;-<span class="st"> </span>copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/runPhylo.html">runPhylo</a></span>(breast_tumor)</a></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="co">## Creating neighbor-joining tree.</span></a>
<a class="sourceLine" id="cb26-2" data-line-number="2"></a>
<a class="sourceLine" id="cb26-3" data-line-number="3"><span class="co">## Access slot with copykit::phylo(scCNA).</span></a>
<a class="sourceLine" id="cb26-4" data-line-number="4"></a>
<a class="sourceLine" id="cb26-5" data-line-number="5"><span class="co">## Done.</span></a></code></pre></div>
<p>Which can be easily visualized with</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="kw">phylo</span>(breast_tumor), <span class="dt">cex =</span> <span class="fl">0.1</span>)</a></code></pre></div>
<p><img src="copykit_workflow_files/figure-gfm/unnamed-chunk-11-1.png"><!-- --></p>
</div>
<div id="plotting-heatmap" class="section level1">
<h1 class="hasAnchor">
<a href="#plotting-heatmap" class="anchor"></a>Plotting heatmap</h1>
<p>Plotting heatmap with <strong>CopyKit</strong> is done with the <code><a href="reference/plotHeatmap.html">plotHeatmap()</a></code> function. By default <code><a href="reference/plotHeatmap.html">plotHeatmap()</a></code> orders the cells according to a neighbor joining phylogenetic tree. <code><a href="reference/plotHeatmap.html">plotHeatmap()</a></code> also checks for the existence of cluster information and will automatically add it as an annotation to the heatmap if available.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1">copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/plotHeatmap.html">plotHeatmap</a></span>(breast_tumor)</a></code></pre></div>
</div>
<div id="visualize-specific-genes" class="section level1">
<h1 class="hasAnchor">
<a href="#visualize-specific-genes" class="anchor"></a>Visualize specific genes</h1>
<p>It might be of interest to check for the heterogeneity in the number of copies across all the samples for different genes. <strong>CopyKit</strong> provides the function <code><a href="reference/geneCopyPlot.html">geneCopyPlot()</a></code>.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1">copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/geneCopyPlot.html">geneCopyPlot</a></span>(breast_tumor,</a>
<a class="sourceLine" id="cb29-2" data-line-number="2">                      <span class="dt">genes =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"KRAS"</span>,</a>
<a class="sourceLine" id="cb29-3" data-line-number="3">                                <span class="st">"MYC"</span>,</a>
<a class="sourceLine" id="cb29-4" data-line-number="4">                                <span class="st">"TP53"</span>,</a>
<a class="sourceLine" id="cb29-5" data-line-number="5">                                <span class="st">"ERBB2"</span>))</a></code></pre></div>
<p><img src="copykit_workflow_files/figure-gfm/unnamed-chunk-13-1.png"><!-- --></p>
</div>


  </div>

</div>



      <footer>
      <div class="copyright">
  <p>Developed by Darlan Conterno Minussi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


