<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CopyKit workflow • CopyKit</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CopyKit workflow">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-invert navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CopyKit</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9003</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/tutorial.html">CopyKit workflow</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>CopyKit workflow</h1>
            
      
      
      <div class="hidden name"><code>tutorial.Rmd</code></div>

    </div>

    
    
<p><strong>CopyKit</strong> provides a toolkit for the analysis of single-cell copy number datasets. It includes functions to read data from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4417119/">Varbin</a> or <a href="https://www.10xgenomics.com/solutions/single-cell-cnv/">10X CNA datasets</a>.</p>
<p>A common workflow with <strong>CopyKit</strong> consists in reading the dataset and using to filter the noisy cells out, clustering and plotting heatmaps for further analysis.</p>
<div id="reading-varbin-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-varbin-dataset" class="anchor"></a>Reading Varbin dataset</h2>
<p>Datasets generated with the varbin pipeline can be read with the function <code><a href="../reference/readVarbinCNA.html">readVarbinCNA()</a></code>. It creates an S4 class <code>scCNA</code> from the output directory of the copy number pipeline. <code><a href="../reference/readVarbinCNA.html">readVarbinCNA()</a></code> searches for the uber*.seg uber.bin and uber.ratio files in the provided directory.</p>
<p>The scCNA object will contain 3 different assays: <strong>segment ratios</strong>, <strong>ratios</strong> and <strong>bincounts</strong> where each bin is row and each sample (cell) is a column.</p>
<p>CopyKit stores the genomic ranges as a GRanges object containing chromosome number, start coordinate, end coordinate and absolute genomic position. Each row represents the coordinates for one bin.</p>
<p>It is possible to exclude the information from ChrY (important when working with XX samples) by setting the argument <code>remove_Y</code> to <code>TRUE</code>. Ex: <code><a href="../reference/readVarbinCNA.html">readVarbinCNA("~/path_to_varbin_output/", remove_Y = TRUE)</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">breast_tumor &lt;-<span class="st"> </span>copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/readVarbinCNA.html">readVarbinCNA</a></span>(<span class="st">"~/path_to_varbin_output/"</span>)</a></code></pre></div>
<pre><code>## ── Attaching packages ───────────────────────────────────────────────────────────────────── tidyverse 1.3.0 ──</code></pre>
<pre><code>## ✓ ggplot2 3.2.1     ✓ purrr   0.3.3
## ✓ tibble  2.1.3     ✓ dplyr   0.8.3
## ✓ tidyr   1.0.0     ✓ stringr 1.4.0
## ✓ readr   1.3.1     ✓ forcats 0.4.0</code></pre>
<pre><code>## ── Conflicts ──────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
## x dplyr::filter() masks stats::filter()
## x dplyr::lag()    masks stats::lag()</code></pre>
<pre><code>## Importing segment ratios.</code></pre>
<pre><code>## Importing ratios.</code></pre>
<pre><code>## Importing bin counts.</code></pre>
<pre><code>## Removed ChrY information.</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">breast_tumor</a></code></pre></div>
<pre><code>## class: scCNA 
## dim: 12167 1442 
## metadata(0):
## assays(3): segment_ratios ratios bin_counts
## rownames(12167): 1 2 ... 12166 12167
## rowData names(1): abspos
## colnames(1442): tn20_2_s2_c10_s394_r1_001 tn20_2_s2_c11_s395_r1_001 ...
##   tn20_s6_c382_s2302_r1_001 tn20_s6_c383_s2303_r1_001
## colData names(1): sample
## reducedDimNames(0):
## spikeNames(0):
## rowRanges has: 12167 ranges
## Phylo:</code></pre>
</div>
<div id="filtering-cells-" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-cells-" class="anchor"></a>Filtering cells.</h2>
<p>Cells with low quality can be easily filtered with <code><a href="../reference/filterCells.html">filterCells()</a></code>.</p>
<p><code><a href="../reference/filterCells.html">filterCells()</a></code> uses a k-nearest-neighbor approach to remove cells with random CNA profiles, largely due to noise.</p>
<p>By default, <strong>CopyKit</strong> automatically outputs a heatmap so you can visually assess if the filtering performed as expected, this however can be turned off by adding the argument <code><a href="../reference/filterCells.html">copykit::filterCells(scCNA, plot_heatmap = FALSE)</a></code>. Plotting the heatmap is a time consuming step due to distance calculations. CopyKit also allows for parallel processing of the distance calculations <code><a href="../reference/filterCells.html">copykit::filterCells()</a></code> with the argument <code>n_threads</code>.</p>
<p>To control how strict the filtering, change the argument <code>resolution</code>. Ex: <code><a href="../reference/filterCells.html">copykit::filterCells(resolution = 0.9)</a></code></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">breast_tumor &lt;-<span class="st"> </span>copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/filterCells.html">filterCells</a></span>(breast_tumor,</a>
<a class="sourceLine" id="cb11-2" data-line-number="2">                            <span class="dt">resolution =</span> <span class="fl">0.8</span>)</a></code></pre></div>
<pre><code>## Calculating correlation matrix.</code></pre>
<pre><code>## Adding information to metadata. Access with SummarizedExperiment::colData(scCNA).</code></pre>
<pre><code>## Plotting heatmap.</code></pre>
<pre><code>## Your dataset has: 1442 Cells. Plotting heatmap may take a long time with large number of cells. Set number of threads with n_threads for parallel processing if possible to speed up.</code></pre>
<pre><code>## Done.</code></pre>
<p><img src="tutorial_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>The information is store as metadata and can be accessed with <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">SummarizedExperiment::colData()</a></code>.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(SummarizedExperiment<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span>(breast_tumor))</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 2 columns
##                                              sample    filtered
##                                         &lt;character&gt; &lt;character&gt;
## tn20_2_s2_c10_s394_r1_001 tn20_2_s2_c10_s394_r1_001        kept
## tn20_2_s2_c11_s395_r1_001 tn20_2_s2_c11_s395_r1_001        kept
## tn20_2_s2_c12_s396_r1_001 tn20_2_s2_c12_s396_r1_001     removed
## tn20_2_s2_c13_s397_r1_001 tn20_2_s2_c13_s397_r1_001     removed
## tn20_2_s2_c14_s398_r1_001 tn20_2_s2_c14_s398_r1_001     removed
## tn20_2_s2_c19_s403_r1_001 tn20_2_s2_c19_s403_r1_001        kept</code></pre>
<p>If you are satisfied with the filtering you can simply subset the object based on the <code>SummarizedExperiment::colData(bt)$filtered</code> column</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">breast_tumor &lt;-<span class="st"> </span>breast_tumor[,SummarizedExperiment<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span>(breast_tumor)<span class="op">$</span>filtered <span class="op">==</span><span class="st"> "kept"</span>]</a></code></pre></div>
<p>After subsetting it is recommended to calculate a distance matrix that will be later passaged to functions such as <code><a href="../reference/runPhylo.html">copykit::runPhylo()</a></code> or <code><a href="../reference/plotHeatmap.html">copykit::plotHeatmap()</a></code>. This distance matrix is stored in the scCNA object and can be accessed with <code>copykit::distMat()</code>.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">breast_tumor &lt;-<span class="st"> </span>copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/runDistMat.html">runDistMat</a></span>(breast_tumor)</a></code></pre></div>
<pre><code>## Calculating distance matrix with metric: euclidean</code></pre>
<pre><code>## Using 48 cores.</code></pre>
<pre><code>## Access distance matrix with copykit::distMat()</code></pre>
<pre><code>## Done.</code></pre>
</div>
<div id="non-linear-dimension-reduction" class="section level2">
<h2 class="hasAnchor">
<a href="#non-linear-dimension-reduction" class="anchor"></a>Non-linear dimension reduction</h2>
<p>We can explore the sample organization at reduced dimensions with <code><a href="../reference/runUmap.html">runUmap()</a></code>. The goal is to explore the dataset for the presence of different subpopulations. <code><a href="../reference/runUmap.html">runUmap()</a></code> stores the generated embedding into the <code>scCNA</code> object and can be accessed with <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SCE-reducedDims.html">SingleCellExperiment::reducedDim(scCNA, 'umap', withDimnames = FALSE)</a></code>. <strong>CopyKit</strong> uses the package <code>uwot</code> to generate the umap embedding, this means that additional parameters that control the global or local structure to <code><a href="../reference/runUmap.html">runUmap()</a></code> can be passed on with the <code>...</code> argument. Ex: <code><a href="../reference/runUmap.html">copykit::runUmap(scCNA, min_dist = 0)</a></code>. To learn more on how those parameters can control global and local structure embedding visit the <a href="https://umap-learn.readthedocs.io/en/latest/clustering.html">Umap website</a>. For the full list of possible parameters check <code><a href="https://rdrr.io/pkg/uwot/man/umap.html">?uwot::umap</a></code>.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1">breast_tumor &lt;-<span class="st"> </span>copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/runUmap.html">runUmap</a></span>(breast_tumor,</a>
<a class="sourceLine" id="cb25-2" data-line-number="2">                                 <span class="dt">min_dist =</span> <span class="dv">0</span>,</a>
<a class="sourceLine" id="cb25-3" data-line-number="3">                                 <span class="dt">n_neighbors =</span> <span class="dv">30</span>)</a></code></pre></div>
<pre><code>## Embedding data with UMAP. Using seed 17</code></pre>
<pre><code>## Access reduced dimensions slot with: SingleCellExperiment::reducedDims(scCNA, 'umap').</code></pre>
<pre><code>## Done.</code></pre>
<p>It is possivel to visualize the results with the function <code><a href="../reference/plotUmap.html">copykit::plotUmap()</a></code>.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1">copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/plotUmap.html">plotUmap</a></span>(breast_tumor)</a></code></pre></div>
<pre><code>## No cluster information detected, use findClusters() to create it.</code></pre>
<pre><code>## Plotting Umap.</code></pre>
<p><img src="tutorial_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div id="finding-clusters" class="section level1">
<h1 class="hasAnchor">
<a href="#finding-clusters" class="anchor"></a>Finding clusters</h1>
<p><strong>CopyKit</strong> uses a graph based approacht to search for clusters in the scCNA dataset. <code><a href="../reference/findClusters.html">copykit::findClusters()</a></code> builds an SNN graph of the k-nearest neighbors and attempts to find two different level of clustering: Major and minor subpopulations. Major clusters are found by looking at the graph connected components, whereas minor clusters uses the Leiden algorithm to detect connected communities within the major clusters. To run <code><a href="../reference/findClusters.html">copykit::findClusters()</a></code> you first need to run <code><a href="../reference/runUmap.html">copykit::runUmap()</a></code> since <strong>CopyKit</strong> applies the clustering on top of the UMAP embedding.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1">breast_tumor &lt;-<span class="st"> </span>copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/findClusters.html">findClusters</a></span>(breast_tumor)</a></code></pre></div>
<pre><code>## Building SNN graph.</code></pre>
<pre><code>## Finding clusters.</code></pre>
<pre><code>## Done.</code></pre>
<p>Clustering information is stored as metadata and can be accessed with `SummarizedExperiment::colData(scCNA).</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(SummarizedExperiment<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span>(breast_tumor))</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 4 columns
##                                              sample    filtered major_clusters
##                                         &lt;character&gt; &lt;character&gt;    &lt;character&gt;
## tn20_2_s2_c10_s394_r1_001 tn20_2_s2_c10_s394_r1_001        kept              A
## tn20_2_s2_c11_s395_r1_001 tn20_2_s2_c11_s395_r1_001        kept              A
## tn20_2_s2_c19_s403_r1_001 tn20_2_s2_c19_s403_r1_001        kept              A
## tn20_2_s2_c2_s386_r1_001   tn20_2_s2_c2_s386_r1_001        kept              B
## tn20_2_s2_c20_s404_r1_001 tn20_2_s2_c20_s404_r1_001        kept              B
## tn20_2_s2_c21_s405_r1_001 tn20_2_s2_c21_s405_r1_001        kept              A
##                           minor_clusters
##                                &lt;numeric&gt;
## tn20_2_s2_c10_s394_r1_001              3
## tn20_2_s2_c11_s395_r1_001              3
## tn20_2_s2_c19_s403_r1_001             12
## tn20_2_s2_c2_s386_r1_001               8
## tn20_2_s2_c20_s404_r1_001              8
## tn20_2_s2_c21_s405_r1_001              3</code></pre>
<p><code><a href="../reference/plotUmap.html">copykit::plotUmap()</a></code> can look for the cluster information and display it.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1">copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/plotUmap.html">plotUmap</a></span>(breast_tumor)</a></code></pre></div>
<pre><code>## Plotting Umap.</code></pre>
<pre><code>## Using colData(scCNA) cluster information.</code></pre>
<p><img src="tutorial_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
</div>
<div id="running-phylogenetic-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#running-phylogenetic-analysis" class="anchor"></a>Running phylogenetic analysis</h1>
<p>It is possible to run phylogenetic analysis with copykit by using the function <code><a href="../reference/runPhylo.html">copykit::runPhylo()</a></code>. This function creates a distance based neighbor-joining tree that is stored within the scCNA object. You can access the resulting tree with <code>copykit::phylo()</code></p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" data-line-number="1">breast_tumor &lt;-<span class="st"> </span>copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/runPhylo.html">runPhylo</a></span>(breast_tumor)</a></code></pre></div>
<pre><code>## Your dataset has: 1081 Cells</code></pre>
<pre><code>## Calculating the distance matrix may take a long time with large number of cells</code></pre>
<pre><code>## Set number of threads with n_threads for parallel processing if possible to speed up.</code></pre>
<pre><code>## Creating neighbor-joining tree.</code></pre>
<pre><code>## Access slot with copykit::phylo(scCNA).</code></pre>
<pre><code>## Done.</code></pre>
<p>Which can be easily visualized with</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb48-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(copykit<span class="op">::</span><span class="kw">phylo</span>(breast_tumor), <span class="dt">cex =</span> <span class="fl">0.1</span>)</a></code></pre></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
</div>
<div id="plotting-heatmap" class="section level1">
<h1 class="hasAnchor">
<a href="#plotting-heatmap" class="anchor"></a>Plotting heatmap</h1>
<p>Plotting heatmap with <strong>CopyKit</strong> is done with the <code><a href="../reference/plotHeatmap.html">plotHeatmap()</a></code> function. By default <code><a href="../reference/plotHeatmap.html">plotHeatmap()</a></code> orders the cells according to a neighbor joining phylogenetic tree. <code><a href="../reference/plotHeatmap.html">plotHeatmap()</a></code> also checks for the existence of cluster information and will automatically add it as an annotation to the heatmap.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" data-line-number="1">copykit<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/copykit/man/plotHeatmap.html">plotHeatmap</a></span>(breast_tumor)</a></code></pre></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#reading-varbin-dataset">Reading Varbin dataset</a></li>
      <li><a href="#filtering-cells-">Filtering cells.</a></li>
      <li><a href="#non-linear-dimension-reduction">Non-linear dimension reduction</a></li>
      <li><a href="#finding-clusters">Finding clusters</a></li>
      <li><a href="#running-phylogenetic-analysis">Running phylogenetic analysis</a></li>
      <li><a href="#plotting-heatmap">Plotting heatmap</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Darlan Conterno Minussi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
